### PR TITLE
perf(inductor): improve `Adam` compile times by shortcutting for loops (via `has_complex`)

### PR DESCRIPTION
**Pull request by jon-chuang**
Adam part of: https://github.com/pytorch/pytorch/issues/110506

TODO:
- If this approach is validated as a good one, it an also be applied to all other optimizers which convert `complex` via list comprehensions

### Results:
`NUM_PARAMS=200, foreach=True`
- main: dynamo: 43s, inductor: 31s, total: 74s
- this PR: dynamo: 3.5s, inductor: 30s, total: 34s (dynamo speedup: 12.3x, overall speedup: 34s, 2.1x)

`NUM_PARAMS=1000, foreach=True, has_complex shortcut`:

```
<class 'torch.optim.adam.Adam'> {'lr': 0.01, 'foreach': True} torch.float32 TorchDynamo compilation metrics:
Function                              Runtimes (s)
------------------------------------  -------------------------------
_compile.<locals>.compile_inner       0.0329, 50.0806, 0.0041
OutputGraph.call_user_compiler        44.9924
```

`NUM_PARAMS=1000, foreach=True, without has_complex shortcut`:
```
<class 'torch.optim.adam.Adam'> {'lr': 0.01, 'foreach': True} torch.float32 TorchDynamo compilation metrics:
Function                              Runtimes (s)
------------------------------------  -------------------------------
_compile.<locals>.compile_inner       0.0389, 58.6069, 0.0043
OutputGraph.call_user_compiler        44.1425
```

### Discussion
- `has_complex` shortcut provides additional 2x dynamo speedup. It is not necessary to achieve a significant overall speedup.

CC: @janeyx99 @mlazos

### CODE REVIEW

File: torch/optim/adam.py [hunk 1]
```diff
@@ -486,11 +494,15 @@ def _multi_tensor_adam(params: List[Tensor],
             device_grads = torch._foreach_neg(device_grads)
 
         # Handle complex parameters
-        device_grads = [torch.view_as_real(x) if torch.is_complex(x) else x for x in device_grads]
-        device_exp_avgs = [torch.view_as_real(x) if torch.is_complex(x) else x for x in device_exp_avgs]
-        device_exp_avg_sqs = [torch.view_as_real(x) if torch.is_complex(x) else x for x in device_exp_avg_sqs]
-        device_max_exp_avg_sqs = [torch.view_as_real(x) if torch.is_complex(x) else x for x in device_max_exp_avg_sqs]
-        device_params = [torch.view_as_real(x) if torch.is_complex(x) else x for x in device_params]
+        # if has_complex:
```
Comment (janeyx99) [inline torch/optim/adam.py:497]: Do you mean to not comment this out?

**Review by janeyx99 — COMMENTED** [review]
I'd want to know the numbers detangling the has_complex heuristic + shortcutting the loops. Your current numbers combines both, yes?

Or is it like this PR and the numbers refer to only the shortcut change?

File: torch/optim/adam.py (hunk 1 above)
Comment (jon-chuang) [reply torch/optim/adam.py:497]: oops, I was running ablation benchmark

**Comment by janeyx99** [comment]
Ah I wasn't sure if the "has_complex" referred to the combination or just the has_complex flag, but I should also get better at reading 😅 A 2X dynamo speedup for num_params=1k seems worth doing, especially when I expect the majority of use cases to not use complex numbers.

**Review by janeyx99 — APPROVED** [review]
Would be good to land after the shortcut lands

File: torch/distributed/optim/functional_adam.py [hunk 2]
```diff
@@ -138,6 +141,7 @@ def step(self, gradients: List[Optional[Tensor]]):
 
         for param, gradient in zip(self.param_group["params"], gradients):
             if gradient is not None:
+                has_complex |= torch.is_complex(param)
```
Comment (janeyx99) [inline torch/distributed/optim/functional_adam.py:144]: This should likely be 
```suggestion
                has_complex |= torch.is_complex(param) || torch.is_complex(gradient)
```
Though I'm curious, @albanD @lezcano is there a known use case where the params and grads differ in complexness?

File: torch/distributed/optim/functional_adam.py [hunk 3]
```diff
@@ -73,6 +73,7 @@ def step_param(self, param: Tensor, grad: Optional[Tensor]):
         exp_avg_sqs = []
         max_exp_avg_sqs = []
         state_steps: List[Tensor] = []
+        has_complex = torch.is_complex(param)
```
Comment (janeyx99) [inline torch/distributed/optim/functional_adam.py:76]: See comment below

File: torch/distributed/optim/functional_adam.py (hunk 2 above)
Comment (albanD) [reply torch/distributed/optim/functional_adam.py:144]: For the stateful version, it is technically impossible for the .grad field to be of different dtype from the Tensor.
You can bypass that with the functional call. But I think it's ok to assume same complex-ness here.

Comment (jon-chuang) [reply torch/distributed/optim/functional_adam.py:144]: Right, I guess we should be worried about functional use-case. 

I think its quite unlikely though, and people will get errors if they try to do this now, so at least it wouldn't go unnnoticed.

Comment (jon-chuang) [reply torch/distributed/optim/functional_adam.py:144]: Anw, my justification for not needing this (which I believe is quite sound): https://github.com/pytorch/pytorch/pull/110613#discussion_r1347853161

In the single tensor case, we only every check `param`

Comment (janeyx99) [reply torch/distributed/optim/functional_adam.py:144]: Ah, then yea, heuristically this is fine to not have. The other comment I put in the other PR about checking is_complex on each state for the functional APIs is still valid though.

Comment (jon-chuang) [reply torch/distributed/optim/functional_adam.py:144]: > checking is_complex on each state for the functional APIs is still valid though.

Hmm could you check again, I'm trying to say that the single_tensor case never does this check, so multi_tensor case should not either.

**Review by lezcano — APPROVED** [review]

**Pull request closed without merging**